File: backend/BeElite/src/main/java/com/salesianos/triana/BeElite/dto/Week/PostWeekDto.java
package com.salesianos.triana.BeElite.dto.Week;

import com.salesianos.triana.BeElite.dto.Program.ProgramDto;
import com.salesianos.triana.BeElite.model.Program;
import com.salesianos.triana.BeElite.model.Week;
import com.salesianos.triana.BeElite.model.Composite_Ids.WeekId;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

public record PostWeekDto(
        @NotEmpty(message = "Week must have name")
        String week_name,
        String description,
        @NotNull(message = "Week must have creation date and time.")
        LocalDateTime created_at,
        @NotNull(message = "Week must be part of a program.")
        ProgramDto program,

        @NotNull(message = "Week must have a valid span.")
        List<LocalDate>span
) {

        public static Week toEntity(PostWeekDto newWeek, UUID p_id, Long weekNumber){
                return Week.builder()
                        .id(WeekId.of(weekNumber, newWeek.week_name, p_id))
                        .createdAt(newWeek.created_at)
                        .description(newWeek.description())
                        .span(newWeek.span)
                        .program(Program.builder().id(p_id).build())
                        .build();
        }
}

File: frontend/be_elite/lib/models/Week/post_week_dto.dart
import 'package:be_elite/models/Program/program_dto.dart';

class PostWeekDto {
  String? weekName;
  String? description;
  String? createdAt;
  ProgramDto? program;
  List<String>? span;

  PostWeekDto({
    this.weekName,
    this.description,
    this.createdAt,
    this.program,
    this.span,
  });

  factory PostWeekDto.fromJson(Map<String, dynamic> json) => PostWeekDto(
        weekName: json['week_name'] as String?,
        description: json['description'] as String?,
        createdAt: json['created_at'] as String?,
        program: json['program'] == null
            ? null
            : ProgramDto.fromJson(json['program'] as Map<String, dynamic>),
        span: json['span'] as List<String>?,
      );

  Map<String, dynamic> toJson() => {
        'week_name': weekName,
        'description': description,
        'created_at': createdAt,
        'program': program?.toJson(),
        'span': span,
      };
}
